### PR TITLE
BBQ Solo Farmer

### PR DESCRIPTION
**Pull request by kichithewolf**
BBQ json: https://github.com/PokemonAutomation/Packages/pull/22
Setup/Instructions: https://github.com/PokemonAutomation/ComputerControl/pull/177

All singleplayer quests except "Pickup 10 items" are automated. That quest will be skipped, no rerolling, it only takes up one slot. No intention of doing multiplayer at this time, I've never found a ditto block in my life.

Some quests more reliable than others, ex. hatch egg and make sandwich are solid but the catch/photo quests are more sensitive to weather-induced lag since they have to navigate the area. I tried aiming for locations that are less likely to lag but there's no avoiding it outright. "Give your Pokemon a bath" depends on what Pokemon it ends up washing, it just moves the brush in a + shape. The program will retry quests anyway, so it will eventually complete the quests. I was able to do a successful run of 100 quests a few times.

Shiny checking - there's checking for overworld encounters that should work in theory, in practice all the targets are fixed spawns so they're full odds and I have no way of really testing it.

There are still improvements that can be done, quest processing logic was written with the assumption that most quests could not be automated so could be tweaked a bit, travel 500 meters could use sound detection (I don't have the sound file for it) to check when it is complete, could probably keep track of egg/TM position, etc. This things is 10 different programs in a trenchcoat, take your time reviewing. I need a break since it's been 3 weeks.

### CODE REVIEW

File: SerialPrograms/Source/PokemonSV/Inference/PokemonSV_BlueberryQuestDetector.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        m_box = ImageFloatBox(0.604, 0.218, 0.282, 0.095);
+        break;
+    case QuestPosition::SECOND:
+        m_box = ImageFloatBox(0.604, 0.395, 0.282, 0.095);
+        break;
+    case QuestPosition::THIRD:
+        m_box = ImageFloatBox(0.604, 0.572, 0.282, 0.095);
+        break;
+    case QuestPosition::FOURTH:
+        m_box = ImageFloatBox(0.604, 0.749, 0.282, 0.096);
```
Comment (pifopi) [inline SerialPrograms/Source/PokemonSV/Inference/PokemonSV_BlueberryQuestDetector.cpp:38]: is the fourth intended to be bigger ? (difference is small anyway)

File: SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    )
+    , OUT_OF_EGGS(
+        "<b>Hatch Quest - Zero Eggs:</b><br>When out of eggs to hatch, do the selected option.",
+        {
+            {OOEggs::Stop,      "stop",         "Stop Program"},
+            {OOEggs::Reroll,    "reroll",       "Reroll Quest - Costs 50BP"},
+            {OOEggs::KeepGoing, "keep-going",   "Keep Going - Warning: Bonus(Red) quests will be blocked!"},
+        },
+        LockMode::LOCK_WHILE_RUNNING,
+        OOEggs::Stop
```
Comment (pifopi) [inline SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.cpp:67]: Any reason not to do KeepGoing ? From my experience, it seems you can "stack up" red quest, meaining even if you do not complete the current red quest, you're already working on unlocking the next red quest

Comment (kichithewolf) [reply SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.cpp:67]: It only stacks up an additional two red quests. The bar fills green->gold->red and then stops there.

image
image

File: SerialPrograms/Source/CommonFramework/OCR/OCR_TextMatcher.cpp [hunk 3]
```diff
@@ -192,9 +192,9 @@ std::vector<uint64_t> binomial_row_u64(size_t degree){
 
     return row;
 }
-uint64_t binomial_coefficient_u64(size_t degree, size_t index){
-    if (degree > 62){
-        throw InternalProgramError(nullptr, PA_CURRENT_FUNCTION, "Cannot go beyond degree 62.");
+double binomial_coefficient_u64(size_t degree, size_t index){
```
[the remark this answers is not in the document]
Comment (kichithewolf) [reply SerialPrograms/Source/CommonFramework/OCR/OCR_TextMatcher.cpp]: renamed it to binomial_coefficient_double

File: SerialPrograms/Source/PokemonSV/Inference/PokemonSV_BlueberryQuestDetector.cpp (hunk 1 above)
Comment (kichithewolf) [reply SerialPrograms/Source/PokemonSV/Inference/PokemonSV_BlueberryQuestDetector.cpp]: fixed, it was unintentional

**Comment by Mysticial** [comment]
Given how big this change is. Let's split this up into smaller PRs that can go in incrementally. Especially the parts that are shared with existing programs so they can be tested separately. I'll comment on each file individually on what should probably be split out.

File: SerialPrograms/Source/CommonFramework/OCR/OCR_TextMatcher.cpp
Comment (Mysticial) [inline SerialPrograms/Source/CommonFramework/OCR/OCR_TextMatcher.cpp:1]: Let's split this binomial refactor out to a separate PR. It's shared by a zillion other things so we can test it with those first.

File: SerialPrograms/Source/PokemonSV/Programs/PokemonSV_Terarium.cpp [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+                if (ret_overworld == 0) {
+                    console.log("Overworld detected.");
+                }
+                context.wait_for_all_requests();
+
+                try {
+                    open_map_from_overworld(info, console, context);
+                }
+                catch (...) {
+                    //Stuck climbing on a wall? Center camera and try to backwards jump off.
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/PokemonSV_Terarium.cpp:69]: The idea of getting off the wall is interesting since we have this problem elsewhere. (platform bot has this problem)

Here you're only retrying once. But from my experience that may not be enough depending on how you get stuck. So what I'm thinking is that we could pull this out in a generic "get off the wall" function with multiple retries before finally giving up.

Here you're using the `open_map_from_overworld()` to determine if you've successfully jumped off the wall. And I think in most of our other cases it's the same thing.

Exactly how we want to do this I'm not sure. Could either add this retry capability to `open_map_from_overworld()` itself or define a new function that wraps it.

File: SerialPrograms/Source/PokemonSV/Programs/PokemonSV_Terarium.cpp
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/PokemonSV_Terarium.cpp:1]: Let's pull this file (and header) out into a separate PR. It looks self-contained enough to test and iterate on in isolation.

Comment (kichithewolf) [reply SerialPrograms/Source/PokemonSV/Programs/PokemonSV_Terarium.cpp:1]: Pulled out into https://github.com/PokemonAutomation/Arduino-Source/pull/409.

File: SerialPrograms/Source/CommonFramework/OCR/OCR_TextMatcher.cpp
Comment (kichithewolf) [reply SerialPrograms/Source/CommonFramework/OCR/OCR_TextMatcher.cpp:1]: Split out into https://github.com/PokemonAutomation/Arduino-Source/pull/410.

**Comment by Mysticial** [comment]
Looks like you can rebase this so we can see what's left to start chipping away at.

**Comment by Mysticial** [comment]
Looks like we need another rebase.

File: SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+#include "PokemonSwSh/Options/PokemonSwSh_BallSelectOption.h"
+#include "PokemonSV/Options/PokemonSV_BattleMoveTable.h"
+#include "PokemonSV/Options/PokemonSV_TeraAIOption.h"
+#include "PokemonSV/Options/PokemonSV_TeraCatchOnWinOption.h"
+
+namespace PokemonAutomation{
+namespace NintendoSwitch{
+namespace PokemonSV{
+
+class BBQOption : public GroupOption, private ConfigOption::Listener {
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.h:25]: Do you need to inherit from `ConfigOption::Listener`? I don't see an override for any of its functions.

File: SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace PokemonAutomation{
+    class BotBaseContext;
+    class ConsoleHandle;
+    struct ProgramInfo;
+namespace NintendoSwitch{
+namespace PokemonSV{
+
+//Navigate to a photo target
+CameraAngle quest_photo_navi(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext& context, BBQOption& BBQ_OPTIONS, BBQuests& current_quest);
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.h:24]: You're not writing to `BBQuests& current_quest` right? If it's strictly an input parameter, take it by value since it's a primitive.

File: SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryQuests.h [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+};
+
+
+// Open the BBQ panel and read the current amount of BP
+int read_BP(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext& context);
+
+//Open the quest panel and read in the current quests.
+//Currently only supports singleplayer.
+//For multiplayer, we will want keep track of which quest are gold/red and scroll cursor down until all of the current player's blue quests are in.
+std::vector<BBQuests> read_quests(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext& context, BBQOption& BBQ_OPTIONS);
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryQuests.h:75]: Can `BBQOption& BBQ_OPTIONS` be made const? I don't think you're changing it. But at the same time, I'm don't remember if the getters are const. (since they internally take locks)

File: SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        true
+    )
+    , BALL_SELECT(
+        "<b>Catch Quest - Ball Select:</b><br>What ball to use when performing a catch quest.",
+        LockMode::UNLOCK_WHILE_RUNNING,
+        "ultra-ball"
+    )
+    , NUM_EGGS(
+        "<b>Hatch Quest - Number of Eggs:</b><br>Amount of eggs located in your current box.",
+        LockMode::LOCK_WHILE_RUNNING, 30, 0, 30
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.cpp:57]: Based on the way you're using this, it's safe to unlock. Though I don't see any legitimate use for editing while running other than fixing an incorrect value.

File: SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        OOEggs::Stop
+    )
+    , FIX_TIME_FOR_HATCH(
+        "<b>Hatch Quest - Fix time for eggs:</b><br>Fix the time before doing a hatch egg quest.",
+        LockMode::UNLOCK_WHILE_RUNNING, false
+    )
+    , CATCH_ON_WIN(true)
+    , NUM_RETRIES(
+        "<b>Number of Retries:</b><br>Reattempt a quest this many times if it fails.",
+        LockMode::LOCK_WHILE_RUNNING, 1, 0, 5
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Options/PokemonSV_BBQOption.cpp:76]: This one looks safe to unlock.

File: SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryQuests.cpp [hunk 10]
```diff
@@ -0,0 +1,1205 @@
...
+    //Put all do-able quests into a different list
+    for (auto n : quest_list) {
+        if (not_possible_quests.contains(n)) {
+            console.log("Quest not possible");
+        }
+        else {
+            //Check eggs remaining.
+            if (n == BBQuests::hatch_egg && eggs_hatched >= BBQ_OPTIONS.NUM_EGGS) {
+                console.log("Out of eggs! Quest not possible.");
+                if (BBQ_OPTIONS.OUT_OF_EGGS == BBQOption::OOEggs::Reroll) {
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryQuests.cpp:231]: If you make this a switch statement so that `BBQ_OPTIONS.OUT_OF_EGGS` is only read once, you can unlock this config.

File: SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.cpp [hunk 11]
```diff
@@ -0,0 +1,1028 @@
...
+                        pbf_mash_button(context, BUTTON_B, 900);
+                        context.wait_for_all_requests();
+                    }
+                    else {
+                        BattleBallReader reader(console, BBQ_OPTIONS.LANGUAGE);
+                        std::string ball_reader = "";
+                        WallClock start = current_time();
+
+                        console.log("Opening ball menu...");
+                        while (ball_reader == "") {
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.cpp:724]: A lot of duplicated code here between the Quick Ball and normal ball paths. Can you pull this out into a separate function where you just pass in the ball to use?

File: SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.cpp [hunk 12]
```diff
@@ -608,6 +610,57 @@ void quest_catch_navi(const ProgramInfo& info, ConsoleHandle& console, BotBaseCo
 
 }
 
+void quest_catch_throw_ball(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext& context, Language language, std::string selected_ball) {
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.cpp:613]: `std::string selected_ball` -> `const std::string& selected_ball` since it's not being modified and is not a primitive.

File: SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.cpp [hunk 13]
```diff
@@ -664,158 +717,71 @@ void quest_catch_handle_battle(const ProgramInfo& info, ConsoleHandle& console,
...
+                        std::chrono::seconds(4),
+                        { battle_menu }
+                    );
+                    if (ret == 0) {
+                        console.log("Battle menu detected early. Using fourth attack.");
+                        MoveSelectWatcher move_watcher(COLOR_BLUE);
+                        MoveSelectDetector move_select(COLOR_BLUE);
+
+                        int ret_move_select = run_until(
+                        console, context,
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.cpp:741]: Indentation

**Comment by Mysticial** [comment]
Last few nits. Otherwise, it's pretty much good to go.

File: SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.h [hunk 14]
```diff
@@ -31,7 +31,7 @@ void quest_photo(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext
 void quest_catch_navi(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext& context, const BBQOption& BBQ_OPTIONS, BBQuests current_quest);
 
 //Select and throw ball
-void quest_catch_throw_ball(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext& context, Language language, std::string selected_ball);
+void quest_catch_throw_ball(const ProgramInfo& info, ConsoleHandle& console, BotBaseContext& context, Language language, const std::string selected_ball);
```
Comment (Mysticial) [inline SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_BlueberryCatchPhoto.h:34]: const reference, not just const. Remember that we don't want to copy non-trivial objects unless we need to.
